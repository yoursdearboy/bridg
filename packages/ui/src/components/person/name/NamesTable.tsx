import { Box, Modal, Table } from "@mantine/core";
import { useDisclosure, useHover } from "@mantine/hooks";
import { IconPencil, IconX } from "@tabler/icons-react";
import type { EntityName } from "api-ts";
import { t } from "i18next";
import { EditNameForm } from "./EditNameForm";
import api from "@/api";

interface NamesTableRowWrapperProps {
  personId: string;
  name: EntityName;
  invalidateQuery: () => void;
}

const NamesTableRowWrapper = ({
  personId,
  name,
  invalidateQuery,
}: NamesTableRowWrapperProps) => {
  const handleDelete = async () => {
    await api.persons.deletePersonsPersonIdNamesNameIdDelete({
      personId,
      nameId: name.id,
    });
    invalidateQuery();
  };

  return (
    <NamesTableRow
      name={name}
      personId={personId}
      onDelete={() => void handleDelete()}
      invalidateQuery={invalidateQuery}
    />
  );
};

interface NamesTableRowProps {
  name: EntityName;
  personId: string;
  invalidateQuery: () => void;
  onDelete: (name: EntityName) => void;
}

const NamesTableRow = ({
  name,
  personId,
  onDelete,
  invalidateQuery,
}: NamesTableRowProps) => {
  const { hovered, ref } = useHover();
  const [opened, { open, close }] = useDisclosure(false);
  const handleEdit = () => {
    open();
  };
  const handleDelete = () => {
    if (window.confirm("Удалить выбранное значение?")) {
      onDelete(name);
    }
  };
  return (
    <>
      <Table.Tr ref={ref}>
        <Table.Td px={0}>{name.label}</Table.Td>
        <Table.Td px={0} style={{ width: 80, display: "flex", gap: 8 }}>
          {hovered && (
            <>
              <IconPencil size={16} color="green" onClick={handleEdit} />
              <IconX size={16} color="red" onClick={handleDelete} />
            </>
          )}
        </Table.Td>
      </Table.Tr>
      <Modal opened={opened} onClose={close} title={t("edit")}>
        <EditNameForm
          personId={personId}
          name={name}
          onCancel={close}
          onSuccess={() => {
            close();
            invalidateQuery();
          }}
        />
      </Modal>
    </>
  );
};

interface NamesTableProps {
  personId: string;
  names: EntityName[];
  invalidateQuery: () => void;
}

export const NamesTable = ({
  personId,
  names,
  invalidateQuery,
}: NamesTableProps) => {
  return (
    <Box pt="md">
      <Table highlightOnHover>
        <Table.Tbody>
          {names.length === 0 ? (
            <Table.Tr>
              <Table.Td px={0} style={{ textAlign: "center" }}>
                {t("nodata")}
              </Table.Td>
            </Table.Tr>
          ) : (
            names.map((name) => (
              <NamesTableRowWrapper
                key={name.id}
                personId={personId}
                name={name}
                invalidateQuery={invalidateQuery}
              />
            ))
          )}
        </Table.Tbody>
      </Table>
    </Box>
  );
};
