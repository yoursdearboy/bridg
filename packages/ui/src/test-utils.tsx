/* eslint-disable react-refresh/only-export-components */
import { MantineProvider } from "@mantine/core";
import { DatesProvider } from "@mantine/dates";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import {
  type AnyRoute,
  createRootRoute,
  createRootRouteWithContext,
  createRoute,
  createRouter,
  RouterProvider,
} from "@tanstack/react-router";
import { act, render } from "@testing-library/react";
import "./i18n";
import type { PropsWithChildren } from "react";

const queryClient = new QueryClient();

const App = ({ children }: React.PropsWithChildren) => {
  return (
    <QueryClientProvider client={queryClient}>
      <MantineProvider defaultColorScheme="auto">
        <DatesProvider settings={{ locale: "en" }}>{children}</DatesProvider>
      </MantineProvider>
    </QueryClientProvider>
  );
};

export const renderRoute = (
  r: AnyRoute,
  { params, props }: { params?: object; props?: PropsWithChildren } = {}
) => {
  const rootRoute = createRootRouteWithContext()();
  let routeConfig = {
    ...r.options,
    params: params,
    getParentRoute: () => rootRoute,
    path: "/",
  };
  if (props?.children)
    routeConfig = { ...routeConfig, component: () => props.children };

  const testRoute = createRoute(routeConfig);
  const routeTree = rootRoute.addChildren([testRoute]);
  const router = createRouter({
    routeTree,
    context: {
      queryClient,
    },
  });
  return act(() =>
    render(
      <App>
        <RouterProvider router={router} />
      </App>
    )
  );
};

export const renderComponent = (component: React.ReactNode) =>
  render(<App>{component}</App>);

export const EmptyRouterProvider = (props: PropsWithChildren) => {
  const rootRoute = createRootRoute({
    component: () => props.children,
  });

  return renderRoute(rootRoute, { props: { children: props.children } });
};
