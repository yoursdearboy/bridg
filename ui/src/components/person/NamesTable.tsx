import api from "@/api";
import {
  Card,
  Stack,
  Group,
  Text,
  Divider,
  Table,
  Button,
} from "@mantine/core";
import {
  QueryClient,
  QueryClientProvider,
  useQuery,
} from "@tanstack/react-query";
import { useTranslation } from "react-i18next";
const queryClient = new QueryClient();

interface NamesTableProps {
  personId: string;
}

export const NamesTable = ({ personId }: NamesTableProps) => {
  const { isPending, error, data } = useQuery({
    queryKey: ["person", personId, "names"],
    queryFn: () =>
      api.persons.indexPersonsPersonIdNamesGet({
        personId,
      }),
  });

  const { t } = useTranslation();

  if (isPending) return t("Loading...");

  if (error) return t("An error has occurred: ") + error.message;

  const rows = data.map((element) => (
    <Table.Tr key={element.use}>
      <Table.Td>{element.family}</Table.Td>
      <Table.Td>{element.given}</Table.Td>
      <Table.Td>{element.middle}</Table.Td>
      <Table.Td>{element.patronymic}</Table.Td>
    </Table.Tr>
  ));

  return (
    <QueryClientProvider client={queryClient}>
      <Card withBorder shadow="sm" radius="md">
        <Card.Section withBorder inheritPadding py="xs">
          <Group justify="space-between">
            <Text fw={500}>{t("Person names")}</Text>

            <Button
              variant="outline"
              fw={500}
              onClick={() => console.log("Add new name")}
            >
              {t("Add Name")}
            </Button>
          </Group>
        </Card.Section>

        <Table>
          <Table.Thead>
            <Table.Tr>
              <Table.Th>{t("Family name")}</Table.Th>
              <Table.Th>{t("Given name")}</Table.Th>
              <Table.Th>{t("Middle name")}</Table.Th>
              <Table.Th>{t("Patronymic")}</Table.Th>
            </Table.Tr>
          </Table.Thead>
          <Table.Tbody>{rows}</Table.Tbody>
        </Table>
      </Card>
    </QueryClientProvider>
  );
};
