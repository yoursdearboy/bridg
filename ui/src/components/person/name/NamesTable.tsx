import { Table } from "@mantine/core";
import type { EntityNameWithId } from "bridg-ts";
import { useTranslation } from "react-i18next";

export const NamesTable = ({ names }: { names: EntityNameWithId[] }) => {
  const { t } = useTranslation();
  const rows = names.map((name) => (
    <Table.Tr key={name.id}>
      <Table.Td>{name.family || "-"}</Table.Td>
      <Table.Td>{name.given || "-"}</Table.Td>
      <Table.Td>{name.middle || "-"}</Table.Td>
      <Table.Td>{name.patronymic || "-"}</Table.Td>
    </Table.Tr>
  ));

  return (
    <>
      <Table striped highlightOnHover>
        <Table.Thead>
          <Table.Tr>
            <Table.Th>{t("Name.family")}</Table.Th>
            <Table.Th>{t("Name.given")}</Table.Th>
            <Table.Th>{t("Name.middle")}</Table.Th>
            <Table.Th>{t("Name.patronymic")}</Table.Th>
          </Table.Tr>
        </Table.Thead>
        <Table.Tbody>
          {names.length === 0 ? (
            <Table.Tr>
              <Table.Td colSpan={4} style={{ textAlign: "center" }}></Table.Td>
            </Table.Tr>
          ) : (
            rows
          )}
        </Table.Tbody>
      </Table>
    </>
  );
};
