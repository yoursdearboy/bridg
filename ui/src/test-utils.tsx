import { MantineProvider } from "@mantine/core";
import { DatesProvider } from "@mantine/dates";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import {
  createRootRoute,
  createRoute,
  createRouter,
  RouterProvider,
  type AnyRoute,
} from "@tanstack/react-router";
import { act, render } from "@testing-library/react";
import "./i18n";

// eslint-disable-next-line react-refresh/only-export-components
const App = ({ children }: React.PropsWithChildren) => {
  const queryClient = new QueryClient();

  return (
    <QueryClientProvider client={queryClient}>
      <MantineProvider defaultColorScheme="auto">
        <DatesProvider settings={{ locale: "en" }}>{children}</DatesProvider>
      </MantineProvider>
    </QueryClientProvider>
  );
};

export const renderRoute = (
  r: AnyRoute,
  { params }: { params?: object } = {}
) => {
  const rootRoute = createRootRoute();
  const testRoute = createRoute({
    ...r.options,
    params: params,
    getParentRoute: () => rootRoute,
    path: "/",
  });
  const routeTree = rootRoute.addChildren([testRoute]);
  const router = createRouter({
    routeTree,
  });
  return act(() =>
    render(
      <App>
        <RouterProvider router={router} />
      </App>
    )
  );
};
