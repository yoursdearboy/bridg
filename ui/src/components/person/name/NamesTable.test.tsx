import { describe, it, expect } from "vitest";

import { renderComponent } from "@/test-utils";
import { NamesTable } from "./NamesTable";

describe("NamesTable", () => {
  it("matches snapshot", () => {
    const { asFragment } = renderComponent(
      <NamesTable
    names={[
          { family: "Charlie", given: "Kai", middle: "May", patronymic: "—" },
          { family: "Nancy", given: "Mia", middle: "Jonson", patronymic: "—" },
        ]}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
