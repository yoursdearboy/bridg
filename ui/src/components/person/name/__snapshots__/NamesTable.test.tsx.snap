// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NamesTable > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <table
    class="m_b23fa0ef mantine-Table-table"
    style="--table-vertical-spacing: calc(0.4375rem * var(--mantine-scale));"
  >
    <thead
      class="m_b242d975 mantine-Table-thead"
    >
      <tr
        class="m_4e7aa4fd mantine-Table-tr"
        data-hover="true"
        data-striped="odd"
        data-with-row-border="true"
      >
        <th
          class="m_4e7aa4f3 mantine-Table-th"
        >
          Family name
        </th>
        <th
          class="m_4e7aa4f3 mantine-Table-th"
        >
          Given name
        </th>
        <th
          class="m_4e7aa4f3 mantine-Table-th"
        >
          Middle name
        </th>
        <th
          class="m_4e7aa4f3 mantine-Table-th"
        >
          Patronymic
        </th>
      </tr>
    </thead>
    <tbody
      class="m_b2404537 mantine-Table-tbody"
    >
      <tr
        class="m_4e7aa4fd mantine-Table-tr"
        data-hover="true"
        data-striped="odd"
        data-with-row-border="true"
      >
        <td
          class="m_4e7aa4ef mantine-Table-td"
        >
          Charlie
        </td>
        <td
          class="m_4e7aa4ef mantine-Table-td"
        >
          Kai
        </td>
        <td
          class="m_4e7aa4ef mantine-Table-td"
        >
          May
        </td>
        <td
          class="m_4e7aa4ef mantine-Table-td"
        >
          —
        </td>
      </tr>
      <tr
        class="m_4e7aa4fd mantine-Table-tr"
        data-hover="true"
        data-striped="odd"
        data-with-row-border="true"
      >
        <td
          class="m_4e7aa4ef mantine-Table-td"
        >
          Nancy
        </td>
        <td
          class="m_4e7aa4ef mantine-Table-td"
        >
          Mia
        </td>
        <td
          class="m_4e7aa4ef mantine-Table-td"
        >
          Jonson
        </td>
        <td
          class="m_4e7aa4ef mantine-Table-td"
        >
          —
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
